### ISSUE TITLE
Handle accessing RequireJS modules without per-file RequireJS wrappers/callbacks

### ISSUE THREAD

**Issue by lrvick** [CaryLandholt/AngularFun#33]
So I have tried everything I can think of and I am kind of at a loss...

I have a requirejs module I want to use within a controller:

https://github.com/faisalman/ua-parser-js/blob/master/src/ua-parser.js#L474-L478

I add it as a dependency of the controller in my main.js shim but in the module itself I am unable to access it directly unless I use a require wrapper/callback which of course breaks things since none of the rest of the angular application is using requirejs wrappers/callbacks.

This works of course, but due to it being in a callback the rest of the controller can't see it on load:

```
require(['ua-parser'], function(UAParser) {
    var parser = new UAParser();
});
```

Ideally I would want the main.js to fully handle the loading in advance like it does with all the angular stuff so that I don't need to use require in my angular code:

```
var parser = new UAParser()
```

Any ideas on a method to resolve cases like these?

**Comment by CaryLandholt** [comment]
Lance,

Whoa - that was a tricky one!

Unlike jQuery (https://github.com/jquery/jquery/blob/master/src/exports.js#L9), ua-parser does not attach to the global window object if AMD (RequireJS) is detected.  They went all-in for AMD, which is good, but a challenge nonetheless.

The trick was continuing to write Angular code without RequireJS and just leverage the shim configuration
So we needed to get UAParser on the window object in order to get a handle to it outside of AMD.

First I created [uaParser.coffee](https://github.com/CaryLandholt/AngularFun/blob/nonAMD/src/scripts/values/uaParser.coffee) for Angular to manage and provide for dependency injection.  I used the value module type - probably should have used a service.  :)

``` CoffeeScript
angular.module('app').value 'uaParser', ->
    # returns a new instance of the UAParser that's on the window object, but it's gotta be there
    new window.UAParser
```

Then we need to make sure the UAParser is on the window object.
We do this in the [shim](https://github.com/CaryLandholt/AngularFun/blob/nonAMD/src/scripts/main.coffee#L18-L21) using the init attribute.

``` CoffeeScript
'values/uaParser':
    deps: ['app', 'libs/ua-parser']
    init: (app, uaParser) ->
        @UAParser = uaParser
```

That's it for wiring it all up.

Now to consume it, we can use it like any other Angular-managed object - inject it.

``` CoffeeScript
angular.module('app').run ['$log', 'uaParser', ($log, uaParser) ->
    $log.info 'uaParser', uaParser()
]
```

I had a branch for nonAMD work, so I stuffed this work in there, even though it really is AMD work.  :)
[Check it out.](https://github.com/CaryLandholt/AngularFun/tree/nonAMD)

Please let me know if this is what you're looking for.  I hope so, because it won't be the last time this scenario comes up.  I'll keep the issue open until you're good to go.

Thanks,
Cary

**Comment by lrvick** [comment]
This seems to work as expected.

In my case I only need this in one file and opted to avoid creating a separate directory/file/server request just for one line of code to use module.value().

Gleaned enough from your example to simply do the following in my shim:

```
'services/MyService':
    deps: ['app', 'libs/ua-parser']
    init: (app, uaParser) ->
        angular.module('app').value('uaParser', -> window.UAParser)
        this.UAParser = new uaParser
```

Then in MyService I can simply import and use uaParser normally with access to all methods by doing

```
var parser = new UAParser()
```

This single-file approach works well for my situation, however the one you outlined is certainly a reasonable general solution.

This was really helpful, thanks!

**Comment by lrvick** [comment]
Actually... seems I jumped the gun.

In either case, the requirement for init-ness outside of angular breaks testacular unit testing.

Unit tests on files that include uaParser die with:

```
    TypeError: 'undefined' is not a constructor (evaluating 'new window.UAParser')
```

**Comment by CaryLandholt** [comment]
Lance,

Keep in mind your code snippet won't work after optimization, since RequireJS is only used in the build process (`grunt prod`).

During optimization the shim file is used to establish the order in which modules are included in the final concatenated and minified file.

The following code will be missing.  This is why I created a separate file for this purpose.

``` CoffeeScript
angular.module('app').value('uaParser', -> window.UAParser)
```

As for testacular (renamed Karma), I found this link which may help.
http://jaketrent.com/post/test-requirejs-testacular/

I'll try to take a look in the next couple days.

Good stuff - thanks Lance!

Cary
